torch/csrc/dynamo/eval_frame.c

```diff
@@ -798,19 +748,84 @@ inline static PyObject* eval_custom_code_impl(
+  // The complexity happens if an argument becomes a cell variable:
+  //  In Python 3.10 and lower, `code.co_cellvars == cell_variables`, and the corresponding slot in args becomes `NULL`.
+  //  In Python 3.11 and higher, `code.co_cellvars > cell_variables`, that cell variable is still stored in args, with a flag set in corresponding item's `co_localspluskinds` .
+  //
+  // ideally, we need to look up new localsplus from old localsplus by name:
+  // for i, name, value in enumerate(localsplusnames_old):
+  //   if value != NULL: (NULL happens for new local variables and arguments that becomes cell variables)
+  //     name_to_idx[name] = i
+  // for i, name in enumerate(localsplusnames_new):
+  //  if name in name_to_idx:
+  //    fastlocals_new[i] = fastlocals_old[name_to_idx[name]]
+  //
+  // The above process of building a `name_to_idx` mapping is expensive.
+  // Dynamo makes the following assumptions:
+  //  1. new code has the same arguments as the old code (both the number and the order)
+  //  2. new code has the same cell variables as the old code (both the number and the order)
+  //  3. new code has the same free variables as the old code (both the number and the order)
+  //  The only flexibility lies in new local variables: new code can introduce their own variables.
+  // With these assumptions, Dynamo can copy data directly by index. Dynamo just needs to take care of copying cell variables correctly.
+  // To avoid runtime cost, the assumptions are checked when we first generate the code object in pytorch/torch/_dynamo/convert_frame.py .
+
+
+  // copy args
+  Py_ssize_t total_argcount_old = frame->f_code->co_argcount + frame->f_code->co_kwonlyargcount + !!(frame->f_code->co_flags & CO_VARARGS) + !!(frame->f_code->co_flags & CO_VARKEYWORDS);
+
+  for (Py_ssize_t i = 0; i < total_argcount_old; i++) {
     Py_XINCREF(fastlocals_old[i]);
     fastlocals_new[i] = fastlocals_old[i];
   }
 
-  for (Py_ssize_t i = 0; i < ncells + nfrees; i++) {
-    Py_XINCREF(fastlocals_old[nlocals_old + i]);
-    fastlocals_new[nlocals_new + i] = fastlocals_old[nlocals_old + i];
+  // copy free vars
+  Py_ssize_t nfrees_old = PyCode_GetNFreevars(frame->f_code);
+
+  for (Py_ssize_t i = 0; i < nfrees_old; i++) {
+    Py_XINCREF(fastlocals_old[n_old - 1 - i]);
+    fastlocals_new[n_new - 1 - i] = fastlocals_old[n_old - 1 - i];
   }
 
+  // copy cell vars, from high index to low index, until it meets a variable that is not cell variable.
+  for (Py_ssize_t i = n_old - nfrees_old - 1, j = n_new - nfrees_old - 1; i >= total_argcount_old; i--, j--) {
+
+  #if IS_PYTHON_3_11_PLUS
+    if(!(_PyLocals_GetKind(frame->f_code->co_localspluskinds, i) & CO_FAST_CELL))
+    {
+      break;
+    }
+  #else
+    if(fastlocals_old[i] == NULL)
```

Comment:
Can you briefly justify in the comments why this is correct?

Reply:
Fixed in [5db055e](https://github.com/pytorch/pytorch/pull/115062/commits/5db055e271f4c1f1d8a3d219d25c142af3ba1788).